Feature/pwa update (#18)
thinktecture/Thinktecture.Blazor: src/Thinktecture.Blazor.PwaUpdate/README.md

```
dotnet add package Thinktecture.Blazor.PwaUpdate
```

## Usage

The package can be used in Blazor WebAssembly projects.

### Add to service collection

To make the UpdateService available on all pages, register it at the IServiceCollection in `Program.cs` before the host is built:

```csharp
builder.Services.AddPwaUpdateServices();
```
### Add to Imports

To use the default UpdateModal component on the hole app razor files, register it in the `_Imports.razor` file.

```html
@using Thinktecture.Blazor.PwaUpdate
```

### Update Modal

To use the default update modal, add the component to the main layout. For example:

```html
@inherits LayoutComponentBase

<UpdateModal InformationMessage="Update available!"></UpdateModal>

...

```
As soon as an update is available, a modal will appear in the upper right corner as shown in the example.

![Pwa Update Model](../../images/update_sample.png)

If you do not want to use the default layout, you can also pass a ChildContent, which will be displayed instead of the default. 

```html
<UpdateModal>
    <div>Update available!</div>
    <button onclick="@Reload">Update</button>
</UpdateModal>
```

> Another option is to override CSS variables:

Author: No, the css come with the package

## Introduction

A Blazor wrapper for the [Service Worker Update](https://web.dev/service-worker-lifecycle/#update-on-reload).

The package allows subscribing to an event that fires as soon as a new update is available.

## Getting started

### Prerequisites

You need .NET 7.0 or newer to use this library.

[Download .NET 7](https://dotnet.microsoft.com/download/dotnet/7.0)

### Platform support

[Platform support for Service Worker](https://caniuse.com/serviceworkers)

### Installation

You can install the package via NuGet with the Package Manager in your IDE or alternatively using the command line:

```
dotnet add package Thinktecture.Blazor.PwaUpdate
```

## Usage

The package can be used in Blazor WebAssembly projects.

### Add to service collection

To make the UpdateService available on all pages, register it at the IServiceCollection in `Program.cs` before the host is built:

```csharp
builder.Services.AddUpdateService();

```
### Add to Imports

To use the default UpdateModal component on the hole app razor files, register it in the `_Imports.razor` file.

```html
@using Thinktecture.Blazor.PwaUpdate
```

### Adapt Publish ServiceWorker

> To get to actual service worker use the `skipWaiting`-method in the service worker. This method forces the waiting service worker to become the active service worker.

Reviewer: ```suggestion
By default, service workers will only update when all tabs of the controlled websites have been closed.
To be able to update an application on reload, you need to call the `skipWaiting()` method of the service worker upon installation.
This skips the service worker's waiting phase, and the new version becomes active instantly.
A reload then leads to the new version of the application the be shown.
Please be aware of the potential implications this may have:
If the application relies on a certain version of the service worker or its cache while it is running, this may lead to errors during runtime.
```

> In Blazor project exists a service worker file `service-worker.published.js`, where a row needs to be added.

Reviewer: ```suggestion
If you want to opt into this behavior, add the following line to the published service worker file `service-worker.published.js`:
```

> ```JavaScript
> async function onInstall(event) {
>     console.info('Service worker: Install');
>     self.skipWaiting(); // --> THIS CALL MUST BE ADDED
> 
>    //...more code
> }

Reviewer: ```suggestion
```diff
async function onInstall(event) {
    console.info('Service worker: Install');
+   self.skipWaiting();
    // … more code
}
```
